`String index out of range` when bulding RPM in Mac

I'm using this plugin into my own "opinionated" build plugin. It works well on Linux, but I just discovered that on Mac I get an exception when trying to build the RPM:
```
> Task :myproject-amzn-rpm:buildRpm
Resolving global dependency management for project 'myproject-amzn-rpm'
Excluding []
Excluding [javax.activation:activation, org.codehaus.jackson:jackson-core-asl, org.codehaus.jackson:jackson-xc, org.jboss.spec.javax.annotation:jboss-annotations-api_1.2_spec, org.codehaus.jackson:jackson-jaxrs, com.sun.jersey:jersey-json, javax.ws.rs:jsr311-api, com.sun.jersey:jersey-core, org.jboss.spec.javax.servlet:jboss-servlet-api_4.0_spec, org.codehaus.jackson:jackson-mapper-asl]
Task ':myproject-amzn-rpm:buildRpm' is not up-to-date because:
  Task has failed previously.
Got an empty spec from org.gradle.api.internal.file.copy.NormalizingCopyActionDecorator$StubbedFileCopyDetails for etc/etc
Got an empty spec from org.gradle.api.internal.file.copy.NormalizingCopyActionDecorator$StubbedFileCopyDetails for etc/rc.d/rc.d
Got an empty spec from org.gradle.api.internal.file.copy.NormalizingCopyActionDecorator$StubbedFileCopyDetails for etc/rc.d/init.d/init.d

> Task :myproject-amzn-rpm:buildRpm FAILED
:myproject-amzn-rpm:buildRpm (Thread[Task worker for ':',5,main]) completed. Took 3.821 secs.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':myproject-amzn-rpm:buildRpm'.
> String index out of range: -89

* Try:
Run with --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':myproject-amzn-rpm:buildRpm'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:110)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.OutputDirectoryCreatingTaskExecuter.execute(OutputDirectoryCreatingTaskExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:101)
        at org.gradle.api.internal.tasks.execution.FinalizeInputFilePropertiesTaskExecuter.execute(FinalizeInputFilePropertiesTaskExecuter.java:44)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:91)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.run(EventFiringTaskExecuter.java:51)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:300)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:292)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:174)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:46)
        at org.gradle.execution.taskgraph.LocalTaskInfoExecutor.execute(LocalTaskInfoExecutor.java:42)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareWorkItemExecutor.execute(DefaultTaskExecutionGraph.java:277)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareWorkItemExecutor.execute(DefaultTaskExecutionGraph.java:262)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:135)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:130)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.execute(DefaultTaskPlanExecutor.java:200)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.executeWithWork(DefaultTaskPlanExecutor.java:191)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.run(DefaultTaskPlanExecutor.java:130)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: -89
        at com.netflix.gradle.plugins.utils.GradleUtils.relativizeSymlink(GradleUtils.groovy:30)
        at com.netflix.gradle.plugins.utils.GradleUtils$relativizeSymlink$0.callStatic(Unknown Source)
        at com.netflix.gradle.plugins.rpm.RpmFileVisitorStrategy.addFile(RpmFileVisitorStrategy.groovy:24)
        at com.netflix.gradle.plugins.rpm.RpmCopyAction.visitFile(RpmCopyAction.groovy:136)
        at com.netflix.gradle.plugins.packaging.AbstractPackagingCopyAction.this$dist$invoke$1(AbstractPackagingCopyAction.groovy)
        at com.netflix.gradle.plugins.packaging.AbstractPackagingCopyAction$StreamAction.methodMissing(AbstractPackagingCopyAction.groovy)
        at com.netflix.gradle.plugins.packaging.AbstractPackagingCopyAction$StreamAction.processFile(AbstractPackagingCopyAction.groovy:59)
        at org.gradle.api.internal.file.copy.NormalizingCopyActionDecorator$1$1.processFile(NormalizingCopyActionDecorator.java:66)
        at org.gradle.api.internal.file.copy.DuplicateHandlingCopyActionDecorator$1$1.processFile(DuplicateHandlingCopyActionDecorator.java:60)
        at org.gradle.api.internal.file.copy.CopyFileVisitorImpl.processFile(CopyFileVisitorImpl.java:62)
        at org.gradle.api.internal.file.copy.CopyFileVisitorImpl.visitFile(CopyFileVisitorImpl.java:46)
        at org.gradle.api.internal.file.AbstractFileTree$FilteredFileTreeImpl$1.visitFile(AbstractFileTree.java:198)
        at org.gradle.api.internal.file.collections.SingletonFileTree.visit(SingletonFileTree.java:42)
        at org.gradle.api.internal.file.collections.FileTreeAdapter.visit(FileTreeAdapter.java:117)
        at org.gradle.api.internal.file.AbstractFileTree$FilteredFileTreeImpl.visit(AbstractFileTree.java:189)
        at org.gradle.api.internal.file.CompositeFileTree.visit(CompositeFileTree.java:93)
        at org.gradle.api.internal.file.copy.CopySpecActionImpl.execute(CopySpecActionImpl.java:39)
        at org.gradle.api.internal.file.copy.CopySpecActionImpl.execute(CopySpecActionImpl.java:24)
        at org.gradle.api.internal.file.copy.DefaultCopySpec$DefaultCopySpecResolver.walk(DefaultCopySpec.java:707)
        at org.gradle.api.internal.file.copy.DefaultCopySpec$DefaultCopySpecResolver.walk(DefaultCopySpec.java:709)
        at org.gradle.api.internal.file.copy.DefaultCopySpec$DefaultCopySpecResolver.walk(DefaultCopySpec.java:709)
        at org.gradle.api.internal.file.copy.DefaultCopySpec.walk(DefaultCopySpec.java:513)
        at org.gradle.api.internal.file.copy.CopySpecBackedCopyActionProcessingStream.process(CopySpecBackedCopyActionProcessingStream.java:38)
        at org.gradle.api.internal.file.copy.DuplicateHandlingCopyActionDecorator$1.process(DuplicateHandlingCopyActionDecorator.java:44)
        at org.gradle.api.internal.file.copy.NormalizingCopyActionDecorator$1.process(NormalizingCopyActionDecorator.java:57)
        at com.netflix.gradle.plugins.packaging.AbstractPackagingCopyAction.execute(AbstractPackagingCopyAction.groovy:43)
        at org.gradle.api.internal.file.copy.NormalizingCopyActionDecorator.execute(NormalizingCopyActionDecorator.java:53)
        at org.gradle.api.internal.file.copy.DuplicateHandlingCopyActionDecorator.execute(DuplicateHandlingCopyActionDecorator.java:42)
        at org.gradle.api.internal.file.copy.CopyActionExecuter.execute(CopyActionExecuter.java:40)
        at org.gradle.api.tasks.AbstractCopyTask.copy(AbstractCopyTask.java:179)
        at com.netflix.gradle.plugins.packaging.SystemPackagingTask.super$5$copy(SystemPackagingTask.groovy)
        at com.netflix.gradle.plugins.packaging.SystemPackagingTask$_copy_closure29.doCall(SystemPackagingTask.groovy:122)
        at com.netflix.gradle.plugins.packaging.SystemPackagingTask$_copy_closure29.doCall(SystemPackagingTask.groovy)
        at com.netflix.gradle.plugins.packaging.SystemPackagingTask.copy(SystemPackagingTask.groovy:121)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:46)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:39)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:26)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:801)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:768)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:131)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:300)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:292)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:174)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:120)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:99)
        ... 31 more

```

Using com.netflix.nebula:gradle-ospackage-plugin:5.2.0, gradle 4.10.2.

The task looks more or less like this:
```
//
            d = /* obtain version data */
            rpm = /* obtain package info and content */
            ospackage {
                packageName rpm.packageName
                release d.buildNumber
                version stripSnapshot(d.version)
                license rpm.license
                vendor rpm.vendor
                url rpm.url
                arch NOARCH
                os LINUX
                prefix rpm.prefix
                summary "${rpm.packageName} (${d.scmInfo})"
                preInstall rpm.preInstallScript()
                preUninstall rpm.preUninstallScript()
                postInstall rpm.postInstallScript()
                postUninstall rpm.postUninstallScript()
                if (rpm.dependencies != null) {
                    rpm.dependencies.each { dep ->
                        requires(dep)
                    }
                }

                if (rpm.genInitScript) {
                    from(rpm.createInitScript()) {
                        into "/etc/rc.d/init.d"
                        user 'root'
                        permissionGroup 'mygroup'
                        fileMode 0755
                    }
                }

                from(configurations.rpmWarDep) {
                    into rpm.warDir
                    user 'root'
                    permissionGroup 'mygroup'
                    fileMode 0440
                }

                from(configurations.rpmJarDep) {
                    into rpm.libDir
                    user 'root'
                    permissionGroup 'mygroup'
                    fileMode 0440
                }

                from('src/main/bin') {
                    into rpm.binDir
                    user 'root'
                    permissionGroup 'mygroup'
                    fileMode 0750
                }

                from(rpm.createVarDir()) {
                    into rpm.varDir
                    user rpm.userName
                    permissionGroup 'mygroup'
                    fileMode 0750
                }

                from(rpm.createConfDir()) {
                    into rpm.confDir
                    user 'root'
                    permissionGroup 'mygroup'
                    fileMode 0750
                }

                from(rpm.createBuildInfo(d)) {
                    into rpm.confDir
                    user 'root'
                    permissionGroup 'mygroup'
                    fileMode 0444
                }


                if (rpm.conflicts != null) {
                    rpm.conflicts.each { x -> conflicts(x) }
                }
            }

```

From the log, it seems it fails early, maybe in the first copy (init script). In there, `rpm.createInitScript()` generates a temporary directory, that contains the init script file that needs to be copied. Could it be that temporary directory differences in Mac (maybe it contains/uses a symlink?) make this fail?

--- a/src/main/groovy/com/netflix/gradle/plugins/utils/GradleUtils.groovy
+++ b/src/main/groovy/com/netflix/gradle/plugins/utils/GradleUtils.groovy
@@ -27,6 +27,10 @@ final class GradleUtils {
     static Tuple2<String, String> relativizeSymlink(FileCopyDetails details, File target) {
         String sourcePath = details.file.path
         String sourceBasePath = sourcePath.substring(0, sourcePath.length() - details.relativeSourcePath.pathString.length())
+        if (!target.path.startsWith(sourceBasePath)) {
+            // target is below sourceBasePath
+            return null
+        }
         String sourceRelative = target.path.substring(sourceBasePath.length())
         String sourceBase = details.path.substring(0, details.path.indexOf(sourceRelative))
 

--- a/src/test/groovy/com/netflix/gradle/plugins/deb/DebPluginTest.groovy
+++ b/src/test/groovy/com/netflix/gradle/plugins/deb/DebPluginTest.groovy
@@ -21,7 +21,6 @@ import nebula.test.ProjectSpec
 import nebula.test.dependencies.DependencyGraph
 import nebula.test.dependencies.GradleDependencyGenerator
 import org.apache.commons.io.FileUtils
-import org.gradle.api.tasks.TaskExecutionException
 import org.redline_rpm.header.Flags
 import spock.lang.Issue
 import spock.lang.Unroll
@@ -1338,4 +1337,31 @@ class DebPluginTest extends ProjectSpec {
         packagedSymlink.isSymbolicLink()
         packagedSymlink.linkName == '../source'
     }
+
+    @Issue("https://github.com/nebula-plugins/gradle-ospackage-plugin/issues/321")
+    def 'resolve symlinks outside build directory'() {
+        File outsideProjectDir = new File(projectDir.parentFile, projectDir.name + "-outside")
+        File source = new File(outsideProjectDir, "source")
+        if (outsideProjectDir.exists()) {
+            outsideProjectDir.deleteDir()
+        }
+        source.mkdirs()
+        def txtFile = new File(source, "test.txt")
+        txtFile.createNewFile()
+        File packageDir = new File(projectDir,"package")
+        java.nio.file.Files.createSymbolicLink(packageDir.toPath(), projectDir.toPath().relativize(outsideProjectDir.toPath()))
+
+        when:
+        project.apply plugin: 'nebula.deb'
+        Deb debTask = project.task([type: Deb], 'buildDeb', {
+            from('package')
+        })
+        debTask.copy()
+
+        then:
+        def scan = new Scanner(debTask.archiveFile.get().asFile)
+        scan.getEntry('./source/').isDirectory()
+        scan.getEntry("./source/test.txt").isFile()
+    }
+
 }

